File: leetcode/cpp/P0062_UniquePaths.cpp
// https://leetcode.com/problems/unique-paths/

#include <array>
#include <iostream>

class Solution {
 public:
  int uniquePaths(int m, int n) {
    std::array<std::array<int, 128>, 128> dp;
    for (int row = 0; row < m; ++row) {
      dp[row].fill(1);
    }

    for (int ii = 1; ii < m; ++ii) {
      for (int jj = 1; jj < n; ++jj) {
        dp[ii][jj] = dp[ii - 1][jj] + dp[ii][jj - 1];
      }
    }
    return dp[m - 1][n - 1];
  }
};

int main(int argc, char** argv) {
  Solution s;
  std::cout << s.uniquePaths(3, 7) << std::endl;
  return 0;
}

File: leetcode/python/P0062_UniquePaths.py
# https://leetcode.com/problems/unique-paths/


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[1 for _ in range(128)] for _ in range(128)]

        for r in range(1, m):
            for c in range(1, n):
                dp[r][c] = dp[r][c - 1] + dp[r - 1][c]

        return dp[m - 1][n - 1]
